File: 05-may/18-jump-game-iv/solution.py
from collections import defaultdict, deque
from typing import List


class Solution:
    def minJumps(self, arr: List[int]) -> int:
        n = len(arr)
        if n == 1:
            return 0

        indices_by_value = defaultdict(list)
        for index, value in enumerate(arr):
            indices_by_value[value].append(index)

        queue = deque([0])
        visited = [False] * n
        visited[0] = True
        steps = 0

        while queue:
            for _ in range(len(queue)):
                index = queue.popleft()
                if index == n - 1:
                    return steps

                neighbors = indices_by_value[arr[index]]
                neighbors.append(index - 1)
                neighbors.append(index + 1)

                for next_index in neighbors:
                    if 0 <= next_index < n and not visited[next_index]:
                        if next_index == n - 1:
                            return steps + 1
                        visited[next_index] = True
                        queue.append(next_index)

                indices_by_value[arr[index]].clear()

            steps += 1

        return -1

File: 05-may/18-jump-game-iv/solution.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Queue;

class Solution {
    public int minJumps(int[] arr) {
        int n = arr.length;
        if (n == 1) {
            return 0;
        }

        Map<Integer, List<Integer>> indicesByValue = new HashMap<>();
        for (int i = 0; i < n; i++) {
            indicesByValue.computeIfAbsent(arr[i], key -> new ArrayList<>()).add(i);
        }

        Queue<Integer> queue = new ArrayDeque<>();
        boolean[] visited = new boolean[n];
        queue.offer(0);
        visited[0] = true;

        int steps = 0;
        while (!queue.isEmpty()) {
            int levelSize = queue.size();

            for (int i = 0; i < levelSize; i++) {
                int index = queue.poll();
                if (index == n - 1) {
                    return steps;
                }

                List<Integer> neighbors = indicesByValue.get(arr[index]);
                neighbors.add(index - 1);
                neighbors.add(index + 1);

                for (int nextIndex : neighbors) {
                    if (nextIndex >= 0 && nextIndex < n && !visited[nextIndex]) {
                        if (nextIndex == n - 1) {
                            return steps + 1;
                        }
                        visited[nextIndex] = true;
                        queue.offer(nextIndex);
                    }
                }

                neighbors.clear();
            }

            steps++;
        }

        return -1;
    }
}
